File: herencia-dto/SergioDeIscar-herencia-java/SergioDeIscar-herencia-java/src/main/java/org/example/controllers/PersonaController.java
package org.example.controllers;

import org.example.models.Persona;
import org.example.repositories.persona.PersonaExtension;
import org.example.repositories.persona.PersonaRepository;

import java.util.List;

public class PersonaController implements PersonaExtension {
    private final PersonaRepository repo;
    public PersonaController(PersonaRepository repo) {
        this.repo = repo;
    }

    @Override
    public List<Persona> getAll() {
        return repo.getAll();
    }

    @Override
    public Persona getById(Integer integer) {
        checkId(integer);
        return repo.getById(integer);
    }

    @Override
    public Persona save(Persona element, boolean storage) {
        checkId(element.getId());
        return repo.save(element, storage);
    }

    @Override
    public void saveAll(List<Persona> elements, boolean storage) {
        elements.forEach(e -> checkId(e.getId()));
        repo.saveAll(elements, storage);
    }

    @Override
    public Persona deleteById(Integer integer) {
        checkId(integer);
        return repo.deleteById(integer);
    }

    @Override
    public Persona update(Persona element) {
        checkId(element.getId());
        return repo.update(element);
    }

    @Override
    public Persona updateById(Integer integer, Persona element) {
        checkId(integer);
        checkId(element.getId());
        return repo.updateById(integer, element);
    }

    private void checkId(Integer id) {
        if (id < 0) throw new IllegalArgumentException("Id no puede ser negativo");
    }
}

File: herencia-dto/SergioDeIscar-herencia-kotlin/SergioDeIscar-herencia-kotlin/src/main/kotlin/controllers/PersonaController.kt
package controllers

import models.Persona
import mu.KotlinLogging
import repositories.persona.PersonaExtension
import repositories.persona.PersonaRepository

private val logger = KotlinLogging.logger {}

class PersonaController(
    private val repo: PersonaRepository
): PersonaExtension {
    override fun getPorcentajePorTipo(): Map<String, Double> {
        logger.debug { "PersonaController ->\tgetPorcentajePorTipo" }
        return repo.getPorcentajePorTipo()
    }

    override fun getAll(): List<Persona> {
        logger.debug { "PersonaController ->\tgetAll" }
        return repo.getAll()
    }

    override fun getById(id: Int): Persona? {
        logger.debug { "PersonaController ->\tgetById" }
        checkId(id)
        return repo.getById(id)
    }

    override fun save(element: Persona, storage: Boolean): Persona {
        logger.debug { "PersonaController ->\tsave" }
        checkId(element.id)
        return repo.save(element, storage)
    }

    override fun saveAll(elements: List<Persona>, storage: Boolean) {
        logger.debug { "PersonaController ->\tsaveAll" }
        elements.forEach { checkId(it.id) }
        repo.saveAll(elements, storage)
    }

    override fun deleteById(id: Int): Persona? {
        logger.debug { "PersonaController ->\tdeleteById" }
        checkId(id)
        return repo.deleteById(id)
    }

    override fun update(element: Persona): Persona? {
        logger.debug { "PersonaController ->\tupdate" }
        checkId(element.id)
        return repo.update(element)
    }

    override fun updateById(id: Int, element: Persona): Persona? {
        logger.debug { "PersonaController ->\tupdateById" }
        checkId(id)
        return repo.updateById(id, element)
    }

    private fun checkId(id: Int) {
        require(id >= 0) { "Id tiene que ser mayor o igual que 0" }
    }
}